pkg/util/http.go

```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
+				return nil, err
+			}
+			userInfo := ""
+			if proxyURL.User != nil {
+				if _, isSet := proxyURL.User.Password(); isSet {
+					userInfo = "*****:*****@"
+				} else {
+					userInfo = "*****@"
+				}
+			}
+
+			log.Debugf("Using proxy %s://%s%s for URL '%s'", proxyURL.Scheme, userInfo, proxyURL.Host, SanitizeURL(r.URL.String()))
+			return proxyURL, nil
+		}
+
+		// no proxy set for this request
+		return nil, nil
+	}
+}
+
+// CreateHTTPTransport creates an *http.Transport for use in the agent
+func CreateHTTPTransport() *http.Transport {
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: config.Datadog.GetBool("skip_ssl_validation"),
+	}
+
+	if config.Datadog.GetBool("force_tls_12") {
+		tlsConfig.MinVersion = tls.VersionTLS12
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	if proxies := config.GetProxies(); proxies != nil {
+		transport.Proxy = GetProxyTransportFunc(proxies)
+	}
+	return transport
+}
+
+// useProxy reports whether requests to addr should use a proxy,
+// according to the noProxy entries, and if a proxy should be used it
+// returns the matching noProxy entry.
+// addr is always a canonicalAddr with a host and port.
+// Copied from `net/http/transport.go`, modified to use the passed noProxy values
+// instead of pulling directly from the no_proxy env var, and to return the
+// matched noProxy entry.
+func useProxy(addr string, noProxy []string) (matches bool, matchingNoProxy string) {
+	if len(addr) == 0 {
+		return true, ""
+	}
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return false, "error splitting host port"
+	}
+	if host == "localhost" {
+		return false, "localhost"
+	}
+	if ip := net.ParseIP(host); ip != nil {
+		if ip.IsLoopback() {
+			return false, "loopback"
+		}
+	}
+
+	addr = strings.ToLower(strings.TrimSpace(addr))
+	if hasPort(addr) {
+		addr = addr[:strings.LastIndex(addr, ":")]
+	}
+
+	for _, p := range noProxy {
+		p = strings.ToLower(strings.TrimSpace(p))
+		matchingNoProxy := p
+		if len(p) == 0 {
+			continue
+		}
+		if p == "*" {
+			return false, matchingNoProxy
+		}
+		if hasPort(p) {
+			p = p[:strings.LastIndex(p, ":")]
+		}
+		if addr == p {
+			return false, matchingNoProxy
+		}
+		if len(p) == 0 {
+			// There is no host part, likely the entry is malformed; ignore.
+			continue
+		}
+		if p[0] == '.' && (strings.HasSuffix(addr, p) || addr == p[1:]) {
+			// no_proxy ".foo.com" matches "bar.foo.com" or "foo.com"
+			return false, matchingNoProxy
+		}
+		if p[0] != '.' && strings.HasSuffix(addr, p) && addr[len(addr)-len(p)-1] == '.' {
+			// no_proxy "foo.com" matches "bar.foo.com"
+			return false, matchingNoProxy
+		}
+	}
+	return true, ""
+}
+
+var portMap = map[string]string{
+	"http":   "80",
+	"https":  "443",
+	"socks5": "1080",
+}
+
+// canonicalAddr returns url.Host but always with a ":port" suffix.
+// Copied from `net/http/transport.go`, modified to remove the idna conversions.
+func canonicalAddr(url *url.URL) string {
```

Comment:
Note: Although to practical effects this code covers all use cases for us, I think it's incomplete strictly speaking. I believe with a SOCKS proxy (which we don't really support), any protocol really could be proxied so we could get a URI for something other than `http`, `https` and then we'd return `foo:0` which would possibly break. I don't think this affects us.

Reply:
thanks for pointing that out, indeed it doesn't affect us so I'll leave it as-is